contracts/modules/capital/CowSwapOperator.sol

```diff
@@ -0,0 +1,240 @@
+      require(sellTokenBalance - totalOutAmount >= swapDetails.minAmount, "SwapOp: swap brings sellToken below min");
+      require(sellTokenBalance - totalOutAmount <= swapDetails.maxAmount, "SwapOp: swap leaves sellToken above max");
+
+      // Ask oracle how much ether we should get
+      uint256 oracleBuyAmount = priceFeedOracle.getEthForAsset(address(order.sellToken), order.sellAmount);
+
+      // Calculate slippage and minimum amount we should accept
+      uint256 maxSlippageAmount = (oracleBuyAmount * swapDetails.maxSlippageRatio) / MAX_SLIPPAGE_DENOMINATOR;
+      uint256 minBuyAmountOnMaxSlippage = oracleBuyAmount - maxSlippageAmount;
+      require(order.buyAmount >= minBuyAmountOnMaxSlippage, "SwapOp: order.buyAmount too low (oracle)");
+
+      // Transfer ERC20 asset from Pool
+      pool.transferAssetToSwapOperator(address(order.sellToken), totalOutAmount);
+
+      // Calculate swapValue using oracle and set it on the pool
+      uint256 swapValue = priceFeedOracle.getEthForAsset(address(order.sellToken), totalOutAmount);
+      pool.setSwapValue(swapValue);
+    } else {
+      revert("SwapOp: Must either sell or buy eth");
+    }
+
+    // Approve Cow's contract to spend sellToken
+    approveVaultRelayer(order.sellToken, totalOutAmount);
+
+    // Store the order UID
+    currentOrderUID = orderUID;
+
+    // Sign the Cow order
+    cowSettlement.setPreSignature(orderUID, true);
+
+    // Emit an event
+    emit OrderPlaced(order);
+  }
+
+  function closeOrder(GPv2Order.Data calldata order) external {
+    // Validate there is an order in place
+    require(currentOrderUID.length > 0, "SwapOp: No order in place");
+
+    // Before validTo, only controller can call this. After it, everyone can call
+    if (block.timestamp <= order.validTo) {
+      require(msg.sender == swapController, "SwapOp: only controller can execute");
+    }
+
+    validateUID(order, currentOrderUID);
+
+    // Check how much of the order was filled, and if it was fully filled
+    uint256 filledAmount = cowSettlement.filledAmount(currentOrderUID);
+    bool fullyFilled = filledAmount == order.sellAmount;
+
+    // Cancel signature and unapprove tokens
+    if (!fullyFilled) {
+      cowSettlement.setPreSignature(currentOrderUID, false);
+      approveVaultRelayer(order.sellToken, 0);
+    }
+
+    // Clear the current order
+    delete currentOrderUID;
+
+    // Withdraw both buyToken and sellToken
+    returnAssetToPool(order.buyToken);
+    returnAssetToPool(order.sellToken);
+
+    // Set swapValue on pool to 0
+    _pool().setSwapValue(0);
+
+    // Emit event
+    emit OrderClosed(order, filledAmount);
+  }
+
+  function returnAssetToPool(IERC20 asset) internal {
+    uint256 balance = asset.balanceOf(address(this));
+
+    if (balance == 0) {
+      return;
+    }
+
+    if (address(asset) == address(weth)) {
+      weth.withdraw(balance); // Unwrap WETH
+      payable(address(_pool())).transfer(balance); // Transfer ETH to pool
```

Comment:
Do not use transfer() for ETH, use the low-level call instead and check the return value.
https://solidity-by-example.org/sending-ether/